jepsen/src/jepsen/core.clj

```diff
@@ -138,72 +138,122 @@
-                                  (= t :info))
-                              (str "Expected client/invoke! to return a map with :type :ok, :fail, or :info, but received " (pr-str completion) " instead")))
-                    (assert (= (:process op) (:process completion)))
-                    (assert (= (:f op)       (:f completion)))
-
-                    ; Log completion
-                    (conj-op! test completion)
-
-                    (if (or (knossos/ok? completion) (knossos/fail? completion))
-                      ; The process is now free to attempt another execution.
-                      process
-                      ; Process hung; move on
-                      (+ process (:concurrency test))))
-
-                  (catch Throwable t
-                    ; At this point all bets are off. If the client or network
-                    ; or DB crashed before doing anything; this operation won't
-                    ; be a part of the history. On the other hand, the DB may
-                    ; have applied this operation and we *don't know* about it;
-                    ; e.g.  because of timeout.
-                    ;
-                    ; This process is effectively hung; it can not initiate a
-                    ; new operation without violating the single-threaded
-                    ; process constraint. We cycle to a new process identifier,
-                    ; and leave the invocation uncompleted in the history.
-                    (conj-op! test (assoc op
-                                          :type :info
-                                          :time  (relative-time-nanos)
-                                          :error (str "indeterminate: "
-                                                      (if (.getCause t)
-                                                        (.. t getCause
-                                                            getMessage)
-                                                        (.getMessage t)))))
-                    (warn t "Process" process "indeterminate")
-                    (+ process (:concurrency test))))))))
-        (info "Worker" process "done")))))
+        (info "Worker" process "with node" node "starting")
+
+        ; Differentiate our base client for this process
+        (let [client (client/open-compat! (:client test) test node)
+              _      (assert client (str "Expected a client, but got " (pr-str client)
+                                         " instead."))
+
+              ; Wait to begin ops until all clients are set up
+              _      (.await setup-barrier)
+
+              ; Consume operations from the generator, returns most recent client when done
+              client (loop [[process client] [process client]]
+                       ; Obtain an operation to execute
+                       (if-let [op (try (generator/op gen test process)
+                                        (catch Exception e e))]
```

Comment:
I think it might be more helpful for users to see a specific error message here, and re-throw (so we don't try to use the exception as an op). Something like `(catch Exception e (warn e "Generator X threw") (throw e))`, which will print out a stacktrace.

Reply:
I agree, thanks! Extracted both cases out into `jepsen.generator/op-and-validate` and re-threw  with additional context.